### ISSUE TITLE
String manipulation corrupts RAM (z80)

### ISSUE THREAD

**Issue by fmaida** [spotlessmind1975/ugbasic#601]
Hello, I'm not sure if this is a bug, a memory corruption or something wrong I wrote...
But when I write this short BASIC program and compile it for the MSX1 target under ugBASIC 1.13.3:

```basic
tt = "MSX ITALIA"
x = 1
dz = 1
DO
	LOCATE x, 11
	s = tt
	IF dz == 1 THEN
		s = " " + s
	ELSE
		s = s + " "
	ENDIF
	PRINT s
	WAIT 150 MS
	x = x + dz
	IF dz == 1 THEN
		IF x > 20 THEN
			x = 20
			dz = -dz
		ENDIF
	ENDIF
	IF dz == -1 THEN
		IF x < 1 THEN
			x = 1
			dz = -dz
		ENDIF
	ENDIF
LOOP
```

The generated ROM will work for a while, then the text will being getting corrupted and finally the machine will reset.
I'm not running the IDE from Windows, I'm using Wine and running it on my Mac. I'm not sure if it is Wine to blame, but so far all the other examples worked as intended.

This is a short video I took from my desktop:

https://user-images.githubusercontent.com/605017/235311366-6663ad4c-2a36-41df-ab77-9afd4fab2c29.mp4

And this is the compiled ROM file for the MSX1:
[prova.msx1.rom.zip](https://github.com/spotlessmind1975/ugbasic/files/11359548/prova.msx1.rom.zip)

**Comment by spotlessmind1975** [comment]
Hi @fmaida , first of all **thank you for the bug reporting!**

I confirm that this is a bug, probably related to string manipulation, and does not depend on Wine. I'm doing some digging to build the simplest case, which causes the problem. At the moment it seems that the problem only occurs with targets equipped with a Zilog Z80 processor.

As a workaround you can change the source as follows:

    x = 1
    dz = 1
    DO
    	LOCATE x, 11
    	IF dz == 1 THEN
    		s = " MSX ITALIA"
    	ELSE
    		s = "MSX ITALIA "
    	ENDIF
    	PRINT s
    	WAIT 150 MS
    	x = x + dz
    	IF dz == 1 THEN
    		IF x > 20 THEN
    			x = 20
    			dz = -dz
    		ENDIF
    	ENDIF
    	IF dz == -1 THEN
    		IF x < 1 THEN
    			x = 1
    			dz = -dz
    		ENDIF
    	ENDIF
    LOOP

**Comment by spotlessmind1975** [comment]
Hello @fmaida , the bug has been fixed on COLDFIX v1.13.3 [rev. 20230430.0047].

If you are using Microsoft Windows and UGBASIC-IDE, please click on **_Configuration_**, **_Compilers_** and download the compiler for `msx1`, in order to fix the issue.

**Comment by fmaida** [comment]
Thank you very much for your quick fix. Greatly appreciated 👍
